### ISSUE TITLE
HTML report collecting fails so master branch is broken

### ISSUE THREAD

**Issue by javiergradiche** [samrocketman/jervis#121]
Hi Sam,

The master branch is broken, jervis_simple branch is working well.

```
hudson.remoting.ProxyException: org.codehaus.groovy.runtime.typehandling.GroovyCastException: Cannot cast object 'build/docs/groovydoc/**/*' with class 'java.lang.String' to class 'java.util.Map'
	at org.codehaus.groovy.runtime.typehandling.DefaultTypeTransformation.continueCastOnSAM(DefaultTypeTransformation.java:405)
	at org.codehaus.groovy.runtime.typehandling.DefaultTypeTransformation.continueCastOnNumber(DefaultTypeTransformation.java:319)
	at org.codehaus.groovy.runtime.typehandling.DefaultTypeTransformation.castToType(DefaultTypeTransformation.java:232)
	at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.castToType(ScriptBytecodeAdapter.java:603)
	at Unknown.Unknown(Unknown)
	at com.cloudbees.groovy.cps.CpsDefaultGroovyMethods.each(CpsDefaultGroovyMethods:2030)
	at com.cloudbees.groovy.cps.CpsDefaultGroovyMethods.each(CpsDefaultGroovyMethods:2015)
	at com.cloudbees.groovy.cps.CpsDefaultGroovyMethods.each(CpsDefaultGroovyMethods:2056)
	at buildViaJervis.buildViaJervisForComponent(/opt/my_jenkins_home/jobs/samrocketman/jobs/jervis/branches/master/builds/1/libs/Jervis Global Library/vars/buildViaJervis.groovy:445)
	at buildViaJervis.call(/opt/my_jenkins_home/jobs/samrocketman/jobs/jervis/branches/master/builds/1/libs/Jervis Global Library/vars/buildViaJervis.groovy:267)
	at ___cps.transform___(Native Method)
	at com.cloudbees.groovy.cps.impl.LocalVariableBlock$LocalVariable.set(LocalVariableBlock.java:45)
	at com.cloudbees.groovy.cps.impl.AssignmentBlock$ContinuationImpl.assignAndDone(AssignmentBlock.java:70)
	at sun.reflect.GeneratedMethodAccessor632.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:72)
	at com.cloudbees.groovy.cps.impl.ContinuationGroup.methodCall(ContinuationGroup.java:60)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.dispatchOrArg(FunctionCallBlock.java:109)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.fixName(FunctionCallBlock.java:77)
	at sun.reflect.GeneratedMethodAccessor640.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:72)
	at com.cloudbees.groovy.cps.impl.ConstantBlock.eval(ConstantBlock.java:21)
	at com.cloudbees.groovy.cps.Next.step(Next.java:83)
	at com.cloudbees.groovy.cps.Continuable$1.call(Continuable.java:174)
	at com.cloudbees.groovy.cps.Continuable$1.call(Continuable.java:163)
	at org.codehaus.groovy.runtime.GroovyCategorySupport$ThreadCategoryInfo.use(GroovyCategorySupport.java:122)
	at org.codehaus.groovy.runtime.GroovyCategorySupport.use(GroovyCategorySupport.java:261)
	at com.cloudbees.groovy.cps.Continuable.run0(Continuable.java:163)
	at org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk(CpsThread.java:174)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:331)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.access$200(CpsThreadGroup.java:82)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:243)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:231)
	at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$2.call(CpsVmExecutorService.java:64)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:131)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Finished: FAILURE
```

**Comment by samrocketman** [comment]
Thanks for the report, I'm aware.  I've been trying to implement publishing HTML reports and that's what is specifically broken.  I've been focusing on other projects and haven't had a chance to circle back.  Let's use this issue to track the bug.

**Comment by samrocketman** [comment]
Started debugging this using the groovy console.  I'll update this comment when I get a reliable means of reproducing this issue in the groovy console.

```groovy
import net.gleske.jervis.lang.lifecycleGenerator
import net.gleske.jervis.lang.pipelineGenerator

def generator = new lifecycleGenerator()
generator.loadLifecyclesString(new File('resources/lifecycles-ubuntu1604-stable.json').text)
generator.loadToolchainsString(new File('resources/toolchains-ubuntu1604-stable.json').text)

generator.loadYamlString('''
language: groovy
jenkins:
  collect:
    html: build/reports/junit
'''.trim())
def pipeline_generator = new pipelineGenerator(generator)
pipeline_generator.supported_collections = ['cobertura', 'junit', 'artifacts', 'html']
pipeline_generator.collect_settings_defaults = [
    html: [
        allowMissing: false,
        alwaysLinkToLastBuild: false,
        includes: '**/*',
        keepAll: false,
        reportFiles: 'index.html',
        reportName: 'HTML Report',
        reportTitles: ''
    ]
]
pipeline_generator.stashmap_preprocessor = [ 
    html: { Map settings ->
        settings['includes']?.tokenize(',').collect {
            "${settings['path']  -~ '/$' -~ '^/'}/${it}"
        }.join(',').toString()
    }
]
pipeline_generator.collect_settings_validation = [
    html: [
        path: '''^[^,\\:*?"'<>|]+$'''
    ]
]

pipeline_generator.stashMap
//pipeline_generator.publishableItems
```

**Comment by samrocketman** [comment]
Master branch is "fixed".  It builds again but HTML report collection is still broken.  I'm going to undertake a major pipeline library refactor effort to make `buildViaJervis()` more readable, separate responsibilities, and make appropriate use of `@NonCPS`.  Hopefully, this will pay off in making it easier for me to fix and extend custom collecting like collecting HTML reports.

**Comment by samrocketman** [comment]
Okay, during refactoring we're back to failing again... heh.  Mainly because it wasn't attempt to collect HTML reports at all before but now it tries again and fails.  I'll keep messing with it.

**Comment by samrocketman** [comment]
Here's a cool screenshot to show master branch is working.

![screenshot from 2018-08-11 21-46-10](https://user-images.githubusercontent.com/875669/43998683-093e4a4a-9db0-11e8-8742-f5bcc7cb7717.png)

This will be generally available when jervis 1.1 is released.

**Comment by samrocketman** [comment]
Summary of problems fixed:

- `pipeline_generator.stashmap_preprocessor` needed to be set from within a `@NonCPS` context.  This was causing the original error message.
- Once I fixed the above point as part of refactoring I realized that getting a `pipeline_generator.stashMap` failed to work even though the preprocessor was working correctly.  There was a minor non-obvious logic error while validing user input on filesets for stashMaps which used the preprocessor.

This effort squashed 2 bugs in the code so I'm satisfied.
